### PR TITLE
Migrate to Tree-Sitter

### LINKED ISSUE #2: Improve declaration intellisense
For the declarations `table`, `elem`, `data`, `type`, `loop`, and `block` we get "unknown" when hovering over labels.
image
This is contrast to function declarations where we get a quick summary of what the label points to.
image
This means when referencing a label later in the file, it's not always obvious what kind of label we're dealing with (if named improperly).

### PR DESCRIPTION
**Pull request by EmNudge**
Turns out tree-sitter has a [pretty slick query language](https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries) which makes matching for identifiers much nicer than the Lezer approach.

It took a bit for me to get it working due to some architecture issues, but I was able to include the wasm build of [tree-sitter-wasm](https://github.com/wasm-lsp/tree-sitter-wasm) (hooray for wasm, again) working fine. This way we don't need to include multiple builds. The repo hasn't be released to npm (and won't, for the foreseeable future), so we're including a build manually. We'll likely need to [add attribution](https://github.com/wasm-lsp/tree-sitter-wasm/blob/2ca28a9f9d709847bf7a3de0942a84e912f59088/LICENSE) once merged.

Fixes https://github.com/NateLevin1/wati/issues/2

## New Features
This is the start of a lot of really interesting features. We can start type-checking functions before compile-time to ensure they return the right kinds of data. We can list regions of used memory. For this change, we're just migrating away from regex for capturing identifiers.

The most notable upgrade here is we're going to start capturing functions/globals/tables on a stack. This way we can retrieve declarations by index. Every identifier is actually just an alias for an index in WAT. 
We can allow both formats of
```wat
(call 0)
(call $main)
```
and provide hover information for both of them.

We have separate "stacks" for locals, globals, funcs, types, tables, labels, and memory. We know which stack to look at based on the context being used. `(br 2)` means the 3rd item in our label stack. `(local.get 1)` means the 2nd item in our local stack (which includes params).

## ToDo
* [x] Add tree sitter wasm
* [x] Add idents collection code
* [x] Replace regex with idents 
* [x] Add import globals
* [ ] Add types?

This PR might not include types. They're not super difficult, but it might hold up the PR from being merged. Every function declaration is also a type declaration which complicates things.

### CODE REVIEW

**Comment by NateLevin1** [comment]
Looking great - I do want to note that for hover info to be useful, we likely will need the type of the variable. Am I understanding this correctly?

**Comment by EmNudge** [comment]
Yeah, technically speaking we're allowed to have a function, a global, a local, and a table with the exact same name. They exist on separate stacks, so their index can align (and thus their aliased name). This is valid:
```wat
(type $main (func (param i32)))
(func $main (type $main) (param $main i32))
(global $main (mut i32) (i32.const 0))
```
We'd need to know in what context an index or variable name is in to know where to go looking for it to retrieve its type.

**Comment by EmNudge** [comment]
Lots of my other projects took the front seat recently, but I'd like to finish up the PR at some point, so here's an update:

This PR now involves completely removing TS in favor of JSDoc and switching to CJS.

## TS → JSDoc

I was really trying to avoid doing this, but our current TS build step includes more than just type-stripping which has made debugging very difficult. You'll see some added code in the build directory. It is migrating from ESM exports to CJS ones. It's doing some backwards compatibility on nullish coalescing. 
Setting breakpoints in a file I didn't build has been difficult for iteration.

The inclusion of regular assets (the wasm file) has also created an infinite loop of TS repeatedly building a recursive directory until your PC runs out of memory. For tree-sitter to even work I'd also need to include a script that copies over the WASM file. This would also mean we'd have redundant assets unless I moved it out into a higher directory.

JSDoc is unfortunately more verbose, but it does fix a lot of our problems there. We completely remove the build step, lowering our final bundle size. We can still use TS Types as long as it's in a separate TS definition file. 

This PR is also going to delete tons more code, so it should get far less messy soon enough.

## ESM → CJS

Surprise surprise, [VS Code has 0 support for ESM in extensions](https://github.com/microsoft/vscode/issues/130367). This one was super frustrating. There's a comment about how to get around all of this, but it completely messes with our TS language server (we're still using `tsconfig.json` here) and is overall a super hacky solution.

Previously we had TS transforming our code as part of the build step. Now we're just going to use require straight up.

### TL;DR

In order to keep working on this, I had to make some larger project changes. Not happy that I had to go in that direction, but everything should come out much cleaner once I finish the PR. I've still yet to delete around 300 more lines of code.

**Comment by NateLevin1** [comment]
As long as you think these changes need to be done, I'm okay with merging them. Everything looks good so far.

**Comment by EmNudge** [comment]
Some updates (long hiatus, again):

I tried to focus on performance a bit too early on. This meant I spent a lot of time dealing with the artifacts of the "fast" system instead of being productive on actual features, which took away some motivation. The focus now is on being **correct** and adding in a caching layer once our feature-set looks refined enough and/or performance tuning becomes important. 

## Current State
The new approach fixes a bunch of problems with the old system. Hovering over an identifier will now have us crawl up the chain to find the source of the identifier. This gives us correct hover information for types, tables, functions, local variables (local + param), and globals. 

Because we're now treating these as separate domains, we don't run into the name-collision issue I mentioned earlier. 

### Possible Perf Impacts
Even though we've pre-computed the tree, we're still running a "query" on every hover. We can cache this for pretty quick wins on subsequent hovers.

We also might want to pre-compute all variable hover information. This isn't terribly difficult, but I worry about massive files (in the megabytes) which people might be opening when inspecting compiled source code.

Regardless, this perf work will be added in a subsequent PR, not this one. The reason being that I'll probably never get to releasing this update if we don't split it off.

## Future Work
This gives us some pretty nice follow-ups:
- Jump to Definition
- Type validation for function arguments

## What's Left For This PR
I want to get this released **soon**. Even in a less-than-perfect state. I imagine the PR right after this will be mostly cleaning up a ton of unused code.

2 features I'm going to add before calling this "done":
- Show information for labels (in br, br_if, block, label)
- Show hover information for **indexes** 
  - I'm excited about this one. It's one of the bigger wins we get with a more structured approach. Variable names are just index aliases and we currently have no support for resolving indexes. This is also something you get when compiling where it removes all symbols. Looking through a WAT source file will show you almost **only** `(call 0)`, `(call_indirect 2)`, etc

**Comment by NateLevin1** [comment]
Wow, this is really great! I've wanted to support indexes for a while and am glad that they'll finally be working.

I agree with you on the performance front - let's not worry too much about it. As long as the initial load into VSCode is not slow, the performance doesn't matter too much.

**Comment by EmNudge** [comment]
And we got the bare-bones migration working!
image

Next PR will probably be jump to definition and cleanup.

**Comment by EmNudge** [comment]
Indexes are live!
image

There are still some small areas I'd like to improve, but as mentioned earlier, I'd like to get this shipped and follow up in a future PR.

**Comment by NateLevin1** [comment]
Looking great, so I'll merge now - here are some little things I noticed that could be addressed in the future. No pressure to implement these, just thought it might be useful 😄: 

- Providing the label of function arguments would be nice - this was actually a feature of the old regex-based parser, so maybe this should be priorized
    - E.g. when hovering over `$test` below, it should show `(func $test (param $someParam i32))` instead of `(func $test (param i32))`
    - ```wat
      (func $test (param $someParam i32))
      ...
      call $test
      ```
    - It would be fine to default to the current shortening behavior if not all of the params have labels
- There's an extra unnecessary parentheses added when hovering non-mut variables - e.g. hovering over `$name` shows `(local $name (i32))`, which has unnecessary parens around `i32`
- Hovering over a variable label in its definition gives "Could not resolve function"
    - This is of questionable utility because it would just be repeating the information right in front of the cursor - but it might be good to fix regardless for consistency
    - For example, try hovering over `$name` below:
      ```wat
      (func $test (param $name i32))
      ```
- Unlabeled indexed branches give "No such label in scope" — maybe they could give the line of the relevant branch? This can be very low priority
    - Example - when hovering over the `0`, it could show "Branch to line #.."
      ```wat
      loop
        br 0
      end
      ```
